src/operator/batch_norm.cc

```diff
@@ -2,41 +2,492 @@
+                    });
+      } else {
+        // when in evaluation mode
+        // Q(X) = X - running_mean  ; i.e. input centered to zero mean
+        // Y = Q(X) / running_std    ; i.e. BN output before weight and bias
+        // dL/dX = w / running_std
+        const AccReal iw = invstd * w;
+        ForEachFast(gradOut, gradIn, static_cast<size_t>(channel),
+                    [iw](const DType *gradOut_data, DType *gradIn_data) {
+                      *gradIn_data = *gradOut_data * iw;
+                    });
+      }
+    }
+
+    // May want to make this a param eventually
+    const AccReal scale = 1.0f;
+
+    if (IsWriting(req[batchnorm::kGamma])) {
+      if (!param_.fix_gamma) {
+        gradWeightData[channel] = scale * dotp * invstd;
+      } else {
+        gradWeightData[channel] = AccReal(0);
+      }
+    }
+
+    if (IsWriting(req[batchnorm::kBeta])) {
+      gradBiasData[channel] = scale * sumGradOut;
+    }
+  }
+}
+
+
 template<>
 Operator *CreateOp<cpu>(BatchNormParam param, int dtype) {
+  Operator *op = nullptr;
 #if MXNET_USE_MKL2017 == 1
-  return new MKLBatchNormOp<cpu, float>(param);
+  if (!param.mkl_off) {
+    switch (dtype) {
+      case mshadow::kFloat32:
+        op = new MKLBatchNormOp<cpu, float>(param);
+        break;
+      case mshadow::kFloat64:
+        op = new MKLBatchNormOp<cpu, double>(param);
+        break;
+      default:
+        // MKL operator doesn't support half_t, so fall through
+        break;
+    }
+  }
+#define BATCHNORM_LOG_MKL_INFO() do { LOG(INFO) << MKLBatchNormOp<cpu, float>::getName() \
+                                          << " Skip MKL optimization"; } while (0)
+#else
+#define BATCHNORM_LOG_MKL_INFO() ((void)0)
 #endif
-  return new BatchNormOp<cpu>(param);
+  if (!op) {
+    MSHADOW_REAL_TYPE_SWITCH_EX(dtype,
+                                DType,
+                                AccReal, {
+                                  BATCHNORM_LOG_MKL_INFO();
+                                  op = new BatchNormOp<cpu, DType, AccReal>(param); });
+  }
+  return op;
 }
 
 // DO_BIND_DISPATCH comes from operator_common.h
 Operator *BatchNormProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_shape,
-    std::vector<int> *in_type) const {
-    std::vector<TShape> out_shape, aux_shape;
-    std::vector<int> out_type, aux_type;
-    CHECK(InferType(in_type, &out_type, &aux_type));
-    CHECK(InferShape(in_shape, &out_shape, &aux_shape));
-    DO_BIND_DISPATCH(CreateOp, param_, (*in_type)[0]);
+                                          std::vector<int> *in_type) const {
+  std::vector<TShape> out_shape, aux_shape;
+  std::vector<int> out_type, aux_type;
+  CHECK(InferType(in_type, &out_type, &aux_type));
+  CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+  BatchNormParam param = param_;
+  CHECK_GE(in_shape->size(), 1U);
+  if ((*in_shape)[0].ndim() != 4) {
+    param.mkl_off = true;
```

Comment:
I don't have access to the shape in CreateOp() in order to determine this. I also figured this was useful for the MKLML problem that we wer ehaving earlier where a certain MKL operator had numerical issues and we had to disable all of MKL rather than having an option to just turn off the one problematic operator (without a recompile).